File: flutter_app/lib/services/auth_service.dart
import 'package:flutter/material.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'dart:convert';
import 'package:dio/dio.dart';

class AuthResponse {
  final String userId;
  final String accessToken;
  final String refreshToken;
  final String username;
  final String displayName;
  final String email;

  AuthResponse({
    this.userId = '',
    this.accessToken = '',
    this.refreshToken = '',
    this.username = '',
    this.displayName = '',
    this.email = '',
  });

  factory AuthResponse.fromJson(Map<String, dynamic> json) => AuthResponse(
        userId: json['userId'] ?? '',
        accessToken: json['accessToken'] ?? '',
        refreshToken: json['refreshToken'] ?? '',
        username: json['username'] ?? '',
        displayName: json['displayName'] ?? '',
        email: json['email'] ?? '',
      );

  Map<String, dynamic> toJson() => {
        'userId': userId,
        'accessToken': accessToken,
        'refreshToken': refreshToken,
        'username': username,
        'displayName': displayName,
        'email': email,
      };
}

class AuthService extends ChangeNotifier {
  static const _storageKey = 'currentUser';
  final _storage = const FlutterSecureStorage();
  final Dio _dio;
  AuthResponse? _currentUser;

  AuthResponse? get currentUser => _currentUser;
  bool get isLoggedIn => _currentUser?.accessToken.isNotEmpty == true;
  String? get token => _currentUser?.accessToken;
  String get userId => _currentUser?.userId ?? '';

  AuthService(this._dio);

  Future<void> init() async {
    final stored = await _storage.read(key: _storageKey);
    if (stored != null) {
      _currentUser = AuthResponse.fromJson(jsonDecode(stored));
      notifyListeners();
    }
  }

  Future<AuthResponse> login(String username, String password) async {
    final res = await _dio.post('/api/auth/login', data: {
      'username': username,
      'password': password,
    });
    final auth = AuthResponse.fromJson(res.data);
    await _setSession(auth);
    return auth;
  }

  Future<AuthResponse> register({
    required String username,
    required String email,
    required String password,
    String? displayName,
    String? invitationCode,
  }) async {
    final res = await _dio.post('/api/auth/register', data: {
      'username': username,
      'email': email,
      'password': password,
      if (displayName != null) 'displayName': displayName,
      if (invitationCode != null) 'invitationCode': invitationCode,
    });
    final auth = AuthResponse.fromJson(res.data);
    await _setSession(auth);
    return auth;
  }

  Future<void> logout() async {
    await _storage.delete(key: _storageKey);
    _currentUser = null;
    notifyListeners();
  }

  Future<void> updateDisplayName(String name) async {
    if (_currentUser != null) {
      _currentUser = AuthResponse(
        userId: _currentUser!.userId,
        accessToken: _currentUser!.accessToken,
        refreshToken: _currentUser!.refreshToken,
        username: _currentUser!.username,
        displayName: name,
        email: _currentUser!.email,
      );
      await _storage.write(
          key: _storageKey, value: jsonEncode(_currentUser!.toJson()));
      notifyListeners();
    }
  }

  Future<void> _setSession(AuthResponse auth) async {
    _currentUser = auth;
    await _storage.write(key: _storageKey, value: jsonEncode(auth.toJson()));
    notifyListeners();
  }
}

File: frontend/src/app/services/auth.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { BehaviorSubject, Observable, tap } from 'rxjs';
import { Router } from '@angular/router';
import { environment } from '../../environments/environment';

export interface AuthResponse {
  userId: string;
  accessToken: string;
  refreshToken: string;
  username: string;
  displayName: string;
  email: string;
  role: string;
}

export interface RegisterRequest {
  username: string;
  email: string;
  password: string;
  displayName?: string;
  invitationCode?: string;
}

export interface LoginRequest {
  username: string;
  password: string;
}

@Injectable({ providedIn: 'root' })
export class AuthService {
  private apiUrl = environment.apiUrl + '/auth';
  private currentUserSubject = new BehaviorSubject<AuthResponse | null>(null);
  currentUser$ = this.currentUserSubject.asObservable();

  constructor(
    private http: HttpClient,
    private router: Router,
  ) {
    const stored = localStorage.getItem('currentUser');
    if (stored) {
      this.currentUserSubject.next(JSON.parse(stored));
    }
  }

  get currentUser(): AuthResponse | null {
    return this.currentUserSubject.value;
  }

  get isLoggedIn(): boolean {
    return !!this.currentUser?.accessToken;
  }

  get token(): string | null {
    return this.currentUser?.accessToken || null;
  }

  register(request: RegisterRequest): Observable<{ message: string; username: string }> {
    return this.http
      .post<{ message: string; username: string }>(`${this.apiUrl}/register`, request);
  }

  login(request: LoginRequest): Observable<AuthResponse> {
    return this.http
      .post<AuthResponse>(`${this.apiUrl}/login`, request)
      .pipe(tap((res) => this.setSession(res)));
  }

  logout(): void {
    localStorage.removeItem('currentUser');
    this.currentUserSubject.next(null);
    this.router.navigate(['/login']);
  }

  private setSession(auth: AuthResponse): void {
    localStorage.setItem('currentUser', JSON.stringify(auth));
    this.currentUserSubject.next(auth);
  }
}
